### PR TITLE
CASSANDRA-18105 3.0

### PR DESCRIPTION
**Pull request by smiklosovic**
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

File: test/distributed/org/apache/cassandra/distributed/test/IndexAndMVDroppingTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+            c.get(1).shutdown(true).get();
+            c.get(1).startup();
+
+            assertEquals(0, c.coordinator(1).execute("SELECT c2 FROM " + KEYSPACE + " .tb", ONE).length);
+        }
+    }
+
+    @Test
+    public void testMaterializedViewDropping() throws Throwable
```
Comment (smiklosovic) [inline test/distributed/org/apache/cassandra/distributed/test/IndexAndMVDroppingTest.java]: I added MV case as well, just to be sure.

File: test/distributed/org/apache/cassandra/distributed/test/IndexAndMVDroppingTest.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+import static org.apache.cassandra.distributed.Cluster.build;
+import static org.apache.cassandra.distributed.api.ConsistencyLevel.ONE;
+import static org.apache.cassandra.distributed.api.Feature.NETWORK;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests CASSANDRA-18105.
+ * <p>
+ * Trucation will insert an entry to system.local into truncated_at map for given table. Because id of an index
```
Comment (Maxwell-Guo) [inline test/distributed/org/apache/cassandra/distributed/test/IndexAndMVDroppingTest.java]: Trucation -> Truncation ?  Am I right?

File: src/java/org/apache/cassandra/db/ColumnFamilyStore.java [hunk 3]
```diff
@@ -538,7 +538,9 @@ public void invalidate(boolean expectMBean)
 
         latencyCalculator.cancel(false);
         compactionStrategyManager.shutdown();
-        SystemKeyspace.removeTruncationRecord(metadata.cfId);
+
+        if (!metadata.isIndex())
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/db/ColumnFamilyStore.java:543]: nit: Might be a good idea to throw a comment inline, something like...

> // Do not remove truncation records for index CFs, given they have the same ID as their backing/base tables.

File: test/distributed/org/apache/cassandra/distributed/test/IndexAndMVDroppingTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+ * is same as id of the base table, when dropping an index, it in fact drops a table with same id as the base table
+ * which will remove an entry for trucated_at column. Hence, after restart of the node, querying the base table
+ * will resurrect the data because commitlog was fully replayed, not taking into consideration (now removed) entry
+ * in truncated_at column.
+ * <p>
+ * The fix consists of not removing an entry in trucated_at column in system.local
+ * if the table being removed is an index.
+ */
+public class IndexAndMVDroppingTest extends TestBaseImpl
+{
```
Comment (maedhroz) [inline test/distributed/org/apache/cassandra/distributed/test/IndexAndMVDroppingTest.java]: nit: We could cut down running time a little if we share the cluster. Something like...

```
private static Cluster CLUSTER;

@BeforeClass
public static void init() throws IOException
{
    CLUSTER = init(Cluster.create(1, conf -> conf.with(NETWORK)));
}

@AfterClass
public static void shutdown()
{
    if (CLUSTER != null)
        CLUSTER.close();
}

@After
public void reset()
{
    if (CLUSTER.get(1).isShutdown())
        CLUSTER.get(1).startup(); // make sure we get the cluster running for test that follow a poorly timed failure
}
```

**Review by maedhroz — APPROVED** [review]
LGTM

Just left a few minor nits

**Review by maedhroz — APPROVED** [review]

**Pull request closed without merging**